# Business Problem

An e-commerce company wants to segment its customers and determine marketing strategies according to these segments.

To this end, we will define the behavior of customers and create groups according to clusters in these behaviors.

In other words, we will include those who exhibit common behaviors in the same groups and we will try to develop special sales and marketing techniques for these groups.

**About DataSet**

https://archive.ics.uci.edu/ml/datasets/Online+Retail+II

The dataset named Online Retail II includes the sales of a UK-based online store between 01/12/2009 - 09/12/2011.

This company sells souvenirs. Think of it like promotional items.

Most of their customers are wholesalers.

**Variables:**

- InvoiceNo: The unique number of each transaction, that is, the invoice. If this code starts with C, it means that the operation has been cancelled.
- StockCode: Product code. Unique number for each product.
- Description: Product name
- Quantity: Number of products. It expresses how many of the products on the invoices have been sold.
- InvoiceDate: Invoice date and time.
- UnitPrice: Product price (GBP)
- CustomerID: Unique customer number
- Country: Country name. Country where the customer lives.

In [1]:
%%html
<style> 
@import url('https://fonts.googleapis.com/css?family=Orbitron|Roboto');
a {color: #37c9e1; font-family: 'Roboto';} 
h1 {color: #C20E69; font-family: 'Poppins'} 
h2, h3 {color: #25B89B; font-family: 'Poppins';}
h4 {color: #818286; font-family: 'Roboto';}
                                      
</style>

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import datetime as dt

#data visualization library
import matplotlib.pyplot as plt

# to display all columns and rows:
pd.set_option('display.max_columns', None); 
pd.set_option('display.max_rows', None);

# the number of digits to be displayed after the comma
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [3]:
# Data in 2010-2011
df_2010_2011 = pd.read_excel("online_retail_II.xlsx", sheet_name = "Year 2010-2011")

In [4]:
# Copying in case need to go back
df = df_2010_2011.copy()

In [5]:
# Descriptive Statistics

def general(dataframe):
    print('Dataset :', '\n')
    print(dataframe.head() ,'\n')
    print("Dataset Shape:" ,dataframe.shape ,'\n')
    print("Columns:" ,'\n')
    print(dataframe.columns, '\n')
    print('Feature Data Types:', '\n')
    print(dataframe.info() ,'\n')
    print("Null Values:", '\n')
    print(dataframe.isnull().sum(),'\n')
    print("Number of Na values in Dataset:", dataframe.isna().any().sum() ,'\n')

general(df)

Dataset : 

  Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom   

Dataset Shape: (541910, 8) 

Columns: 

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='object') 

Feature Data Types: 

<cl

In [6]:
# Number of unique products

df["Description"].nunique()

4223

In [7]:
# Total number of unique products

df["Description"].value_counts().head()

WHITE HANGING HEART T-LIGHT HOLDER    2369
REGENCY CAKESTAND 3 TIER              2200
JUMBO BAG RED RETROSPOT               2159
PARTY BUNTING                         1727
LUNCH BAG RED RETROSPOT               1638
Name: Description, dtype: int64

In [8]:
# Most ordered product (descending)

df.groupby("Description").agg({"Quantity":"sum"}).sort_values("Quantity", ascending = False).head()

Quantity
Description                                
WORLD WAR 2 GLIDERS ASSTD DESIGNS     53847
JUMBO BAG RED RETROSPOT               47363
ASSORTED COLOUR BIRD ORNAMENT         36381
POPCORN HOLDER                        36334
PACK OF 72 RETROSPOT CAKE CASES       36039

In [9]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom

In [10]:
# There is information about the unit price and the number of products purchased(Quantity),
# but there is no information about the total price, let's create a new variable by assigning it.

df["TotalPrice"] = df["Quantity"] * df["Price"]

In [11]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34

In [12]:
# Canceled transactions

df[df["Invoice"].str.contains("C", na = False)].head()

Invoice StockCode                       Description  Quantity  \
141  C536379         D                          Discount        -1   
154  C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235  C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236  C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237  C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   

            InvoiceDate  Price  Customer ID         Country  TotalPrice  
141 2010-12-01 09:41:00  27.50     14527.00  United Kingdom      -27.50  
154 2010-12-01 09:49:00   4.65     15311.00  United Kingdom       -4.65  
235 2010-12-01 10:24:00   1.65     17548.00  United Kingdom      -19.80  
236 2010-12-01 10:24:00   0.29     17548.00  United Kingdom       -6.96  
237 2010-12-01 10:24:00   0.29     17548.00  United Kingdom       -6.96

In [13]:
# Which product has the most returns?

Cdf = df[df["Invoice"].str.contains("C", na=False)]
Cdf["Description"].value_counts().head(10)

Manual                               244
REGENCY CAKESTAND 3 TIER             181
POSTAGE                              126
JAM MAKING SET WITH JARS              87
Discount                              77
SET OF 3 CAKE TINS PANTRY DESIGN      74
SAMPLES                               61
STRAWBERRY CERAMIC TRINKET BOX        55
ROSES REGENCY TEACUP AND SAUCER       54
RECIPE BOX PANTRY YELLOW DESIGN       47
Name: Description, dtype: int64

In [14]:
# Remove returns from dataset

df = df[~df["Invoice"].str.contains("C", na = False)]
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75     17850.00  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34

In [15]:
# Amount of each invoice 

df.groupby("Invoice").agg({"TotalPrice":"sum"}).head()

TotalPrice
Invoice            
536365       139.12
536366        22.20
536367       278.73
536368        70.05
536369        17.85

In [16]:
# Well, how many orders were placed from which country?

df["Country"].value_counts().head()

United Kingdom    487622
Germany             9042
France              8409
EIRE                7894
Spain               2485
Name: Country, dtype: int64

In [17]:
#  Country that makes the most money (descending)

df.groupby("Country").agg({"TotalPrice":"sum"}).sort_values("TotalPrice", ascending= False).head()

TotalPrice
Country                   
United Kingdom  9003097.96
Netherlands      285446.34
EIRE             283453.96
Germany          228867.14
France           209733.11

In [18]:
# Number of Null Values

df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    134697
Country             0
TotalPrice          0
dtype: int64

In [19]:
# Dropping na values

df.dropna(inplace= True)
df.isnull().sum()

Invoice        0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
Price          0
Customer ID    0
Country        0
TotalPrice     0
dtype: int64

In [21]:
# After dropping na values, data set's shape becoming from (541910, 8) to (397925, 9) ** We added TotalPrice column, therefor have 9 observations

df.shape

(397925, 9)

# CUSTOMER SEGMENTATION WITH RFM SCORES

It consists of the initials of the expressions Recency, Frequency, Monetary.

It is a technique that helps determine marketing and sales strategies based on customers' purchasing habits.

# Recency

Recency: The time from the customer's last purchase to date

- In other words, it is “the time elapsed from the last contact of the customer until today”.

- Today's date - Last purchase

- For example, if we're doing this analysis today, today's date - the last product purchase date.

- It could be 20 or 100 for example. We know that the customer who is 20 has had contact with us recently.


In [23]:
df.head(2)

Invoice StockCode                         Description  Quantity  \
0  536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                 WHITE METAL LANTERN         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55     17850.00  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39     17850.00  United Kingdom       20.34

In [24]:
# First invoice date in dataset

df["InvoiceDate"].min()

Timestamp('2010-12-01 08:26:00')

In [25]:
# Last invoice date in dataset

df["InvoiceDate"].max()

Timestamp('2011-12-09 12:50:00')

In [26]:
# Specifying the last invoice date as today's date

today_date = dt.datetime(2011, 12, 9)
today_date

datetime.datetime(2011, 12, 9, 0, 0)

In [27]:
# Each customer's last purchase information

df.groupby("Customer ID").agg({"InvoiceDate":"max"}).head()

InvoiceDate
Customer ID                    
12346.00    2011-01-18 10:01:00
12347.00    2011-12-07 15:52:00
12348.00    2011-09-25 13:13:00
12349.00    2011-11-21 09:51:00
12350.00    2011-02-02 16:01:00

In [28]:
#  Converting Customer ID to int

df["Customer ID"] = df["Customer ID"].astype(int)
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00   2.55        17850  United Kingdom       15.30  
1 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
2 2010-12-01 08:26:00   2.75        17850  United Kingdom       22.00  
3 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34  
4 2010-12-01 08:26:00   3.39        17850  United Kingdom       20.34

In [29]:
# The time from the customer's last purchase to date (Recency)

temp_df = (today_date - df.groupby("Customer ID").agg({"InvoiceDate":"max"}))
temp_df.rename(columns={"InvoiceDate":"Recency"},inplace= True)
temp_df.head()

Recency
Customer ID                  
12346       324 days 13:59:00
12347         1 days 08:08:00
12348        74 days 10:47:00
12349        17 days 14:09:00
12350       309 days 07:59:00

In [30]:
# Getting only day numbers 

recency_df = temp_df["Recency"].apply(lambda x: x.days)
recency_df.head()

Customer ID
12346    324
12347      1
12348     74
12349     17
12350    309
Name: Recency, dtype: int64

In [31]:
# Doing all these steps in one line without creating temp_df:

recency_df = df.groupby("Customer ID").agg({"InvoiceDate": lambda x: (today_date - x.max()).days})
recency_df.rename(columns = {"InvoiceDate":"Recency"}, inplace = True)
recency_df.head()

Recency
Customer ID         
12346            324
12347              1
12348             74
12349             17
12350            309

# Frequency

Frequency: Total number of purchases.

In [32]:
# nunique is used instead of count because there are multiple invoices

freq_df = df.groupby("Customer ID").agg({"InvoiceDate":"nunique"})
freq_df.rename(columns={"InvoiceDate": "Frequency"}, inplace=True)
freq_df.head()

Frequency
Customer ID           
12346                1
12347                7
12348                4
12349                1
12350                1

# Monetary

Monetary: The total spend by the customer.

In [33]:
monetary_df = df.groupby("Customer ID").agg({"TotalPrice":"sum"})
monetary_df.rename(columns={"TotalPrice": "Monetary"}, inplace=True)
monetary_df.head()

Monetary
Customer ID          
12346        77183.60
12347         4310.00
12348         1797.24
12349         1757.55
12350          334.40

In [34]:
# Shape Control

print(recency_df.shape,freq_df.shape,monetary_df.shape)

(4339, 1) (4339, 1) (4339, 1)


# RFM

In [35]:
# RFM table for scoring
rfm = pd.concat([recency_df, freq_df, monetary_df], axis=1)
rfm.head()

Recency  Frequency  Monetary
Customer ID                              
12346            324          1  77183.60
12347              1          7   4310.00
12348             74          4   1797.24
12349             17          1   1757.55
12350            309          1    334.40

## RFM Scoring

The recency variable is when the customer made the last purchase, the smaller the value is more important to us, so the smaller value scores higher.
In Frequency and Monetary variables, high values get higher scores.

In [36]:
# qcut function : cutting data according to quantiles

rfm["RecencyScore"]= pd.qcut(rfm["Recency"], 5, labels= [5,4,3,2,1])
rfm["FrequencyScore"]= pd.qcut(rfm["Frequency"].rank(method="first"), 5, labels= [1,2,3,4,5])
rfm["MonetaryScore"]= pd.qcut(rfm["Monetary"], 5, labels= [1,2,3,4,5])
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          1  77183.60            1              1   
12347              1          7   4310.00            5              5   
12348             74          4   1797.24            2              4   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore  
Customer ID                
12346                   5  
12347                   5  
12348                   4  
12349                   4  
12350                   2

In [37]:
# combining all scores into one variable

rfm["RFM_SCORE"] = (rfm['RecencyScore'].astype(str) + 
                    rfm['FrequencyScore'].astype(str) + 
                    rfm['MonetaryScore'].astype(str))
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          1  77183.60            1              1   
12347              1          7   4310.00            5              5   
12348             74          4   1797.24            2              4   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore RFM_SCORE  
Customer ID                          
12346                   5       115  
12347                   5       555  
12348                   4       244  
12349                   4       414  
12350                   2       112

## RFM Segmentation


Scoring according to last purchase to date (recency) and number of purchases (frequency). 

In [38]:
# RFM Segmentation with Regular Expressions (REGEX)

seg_map = {
    r'[1-2][1-2]': 'Hibernating',
    r'[1-2][3-4]': 'At Risk',
    r'[1-2]5': 'Can\'t Lose',
    r'3[1-2]': 'About to Sleep',
    r'33': 'Need Attention',
    r'[3-4][4-5]': 'Loyal Customers',
    r'41': 'Promising',
    r'51': 'New Customers',
    r'[4-5][2-3]': 'Potential Loyalists',
    r'5[4-5]': 'Champions'
}

In [39]:
rfm['Segments'] = rfm['RecencyScore'].astype(str) + rfm['FrequencyScore'].astype(str)
rfm['Segments'] = rfm['Segments'].replace(seg_map, regex=True)
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          1  77183.60            1              1   
12347              1          7   4310.00            5              5   
12348             74          4   1797.24            2              4   
12349             17          1   1757.55            4              1   
12350            309          1    334.40            1              1   

            MonetaryScore RFM_SCORE     Segments  
Customer ID                                       
12346                   5       115  Hibernating  
12347                   5       555    Champions  
12348                   4       244      At Risk  
12349                   4       414    Promising  
12350                   2       112  Hibernating

In [40]:
#It is now easier to take action regarding our segmented customers. 
# For example, We can advise the relevant department to take action by sending 
# the customer numbers that "Need Attention" segment.

rfm[rfm["Segments"] == "Need Attention"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12360             51          3   2662.06            3              3   
12372             70          3   1298.04            3              3   
12413             65          3    758.10            3              3   
12456             43          3   3181.04            3              3   
12536             42          3  12601.83            3              3   

            MonetaryScore RFM_SCORE        Segments  
Customer ID                                          
12360                   5       335  Need Attention  
12372                   4       334  Need Attention  
12413                   3       333  Need Attention  
12456                   5       335  Need Attention  
12536                   5       335  Need Attention

In [41]:
# Customer ID of Need Attention segment

rfm[rfm["Segments"] == "Need Attention"].index

Int64Index([12360, 12372, 12413, 12456, 12536, 12577, 12759, 12808, 12879,
            12885,
            ...
            17928, 17934, 17999, 18147, 18160, 18205, 18228, 18252, 18261,
            18270],
           dtype='int64', name='Customer ID', length=184)

## Profile of Segments 

We can profile segments using descriptive statistics.
For exemple; we can see the number of customers and statistics for each segment.

In [42]:
rfm[["Segments","Recency","Frequency","Monetary"]].groupby("Segments").agg(["mean","count","max"])

Recency            Frequency            Monetary        \
                       mean count  max      mean count  max     mean count   
Segments                                                                     
About to Sleep        51.37   353   70      1.16   353    2   472.55   353   
At Risk              152.04   594  372      2.87   594    5  1079.81   594   
Can't Lose           130.05    64  371      8.31    64   33  2791.01    64   
Champions              4.37   632   11     12.34   632  211  6866.78   632   
Hibernating          215.66  1069  372      1.10  1069    2   488.86  1069   
Loyal Customers       31.69   820   70      6.44   820   61  2862.89   820   
Need Attention        50.27   184   70      2.32   184    3   894.49   184   
New Customers          5.43    42   11      1.00    42    1   388.21    42   
Potential Loyalists   15.37   486   31      2.01   486    3  1041.34   486   
Promising             21.42    95   31      1.00    95    1   290.91    95   

                               
                          max  
Segments                       
About to Sleep        6207.67  
At Risk              44534.30  
Can't Lose           10254.18  
Champions           280206.02  
Hibernating          77183.60  
Loyal Customers     124914.53  
Need Attention       12601.83  
New Customers         3861.00  
Potential Loyalists 168472.50  
Promising             1757.55

# COMMENTS AND SUGGESTIONS

I will comment on 3 segments based on descriptive statistics.

## Champions

It is our customer segment with the highest frequency of visits and shopping.
In this segment, where 665 people are present, the last shopping took place on average 4 days ago.
we earned an average of 6867 from our customers in this segment.

We can ensure that they stay in this segment by offering campaigns and discounts that will make them feel special according to their individual preferences. 
It should be among our preferences to present new products to this group first and to send e-mails showing that we remember their special days.

## Can't Loose

This segment, which has high shopping frequencies and money returns, has not visited our site for an average of 130 days. 
We should prevent them from falling into the "At Risk" segment by suggesting limited-time campaigns and communicating with customer surveys.

## Need Attention

It is in our hands to pull this segment to loyal customers without falling into the about to sleep segment.
We can increase the frequency of shopping with constant reminders and limited time coupons.

### Special thakns to;
- https://bootcamp.veribilimiokulu.com/egitim/veri-bilimci-yetistirme-programi/
- https://github.com/mvahit